### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by btcdrak**
Add a CONTRIBUTING.md as per https://github.com/blog/1184-contributing-guidelines

This document was written in collaboration with @laanwj.

### CODE REVIEW

**Comment by paveljanik** [comment]
Please add a line or two about Trivial tree?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    [P2P] Add ZMQ message support
+    [Consensus] Add new opcode for BIP-XXXX OP_CHECKAWESOMESIG
+    [Qt Wallet] Refactor send code
+
+If a pull request is specifically not to be considered for merging (yet) please prefix the title with [WIP] or use [Tasks Lists](https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments) in markdown to the body of the pull request to indicate tasks are pending.
+
+The body of the pull request should contain enough description about what the patch does together with any justification/reasoning. You should include references to any discussions (for example other tickets or mailing list discussions).
+
+  - At this stage one should expect comments and review from other contributors. You can add more commits to your pull request by committing them locally and pushing to your fork until you have satisfied all feedback. If your pull request is accepted for merging, you may be asked by a maintainer to squash and or rebase your commits before it will be merged. The length of time for required for peer review is unpredictable and will vary from patch to patch.
```
Comment (MarcoFalke) [inline CONTRIBUTING.md]: Nit: Remove `-`.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+###Peer Review
+
+Anyone may participate in peer review is expressed by comments in the pull request. Typically reviewers will review the code for obvious errors, as well as test out the patch set and opine on the technical merits of the patch. Project maintainers take into account the peer review when determining if there is consensus to merge a pull request (remember that discussions may have been spread out over github, mailing list and IRC discussions). The following language is used within pull-request comments:
+
+  - ACK means "I have tested the code and I agree it should be merged"
+  - NACK means "I disagree this should be merged", and must be accompanied by sound technical justification. NACKs without accompanying reasoning may be disregarded.
+  - utACK means "I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged"
+  - Concept ACK mean "I agree in the general principle of this pull request"
+
```
Comment (MarcoFalke) [inline CONTRIBUTING.md]: Nit: Should "Nit" be included in this list?

**Comment by laanwj** [comment]
ACK, thanks for writing this up

**Comment by paveljanik** [comment]
ACK, very well written!

**Comment by gavinandresen** [comment]
ACK, very nice!

**Comment by laanwj** [comment]
@paveljanik Re: the trivial tree, maybe @theuni could comment on that. I think it was a great idea, but AFAIK it has not been kept up to date regularly lately, so I 'm not sure what his stance on it is now.

**Comment by paveljanik** [comment]
If it is so, I think that adding `[Trivial]` to pull request titles is enough to solve this.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  - Create topic branch
+  - Commit patches 
+
+The project coding conventions in [doc/developer-notes.md](doc/developer-notes.md) must be adhered to.
+
+In general [commits should be atomic](https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention) and diff should be easy to read. For this reason do not mix any formatting fixes or code moves with actual code changes.
+
+Commit messages should be verbose by default consisting of a short subject line (50 chars max), a blank line and detailed explanatory text as separate paragraph(s); unless the title alone is self-explanatory (like "fixed spelling") then a single title line is sufficient. Commit messages should be helpful to people reading your code in the future, so explain the reasoning for your decisions. Further explanation [here](http://chris.beams.io/posts/git-commit/).
+
+If a particular commit references another issue, please add the reference, for example "refs #1234", or "fixes #4321". Using "fixes or closes" keywords will cause the corresponding issue to be closed when the pull request is merged.
```
Comment (jonasschnelli) [inline CONTRIBUTING.md]: I'm not sure if this is a good idea. Git and Github are two separate things. I agree with referencing a fix by `fixes #4321` placed in a *Github* PR description or a *Github* comment. Placing it in a *git commit* is not required and not best practice IMO.

Comment (MarcoFalke) [inline CONTRIBUTING.md]: +1. I prefer to place it somewhere in the GitHub PR description.

**Comment by jonasschnelli** [comment]
ACK.
Nice work!

File: CONTRIBUTING.md (hunk 3 above)
Comment (sipa) [inline CONTRIBUTING.md]: Agree, to the extent possible you should be able to read and understand the
commits without needing access to the github repository or its PRs/issues.

Comment (laanwj) [inline CONTRIBUTING.md]: Absolutely, it must be understandable without referring to them, but please **do** refer to solved github issues in the commit descriptions. This makes it much easier to keep track of what was solved when when browsing history.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+The Bitcoin Core project operates an open contributor model where anyone is welcome to contribute towards development in the form of peer review, testing and patches. This document explains the practical process and guidelines for contributing.
+
+Firstly in terms of structure, there is no particular concept of “Core developers” in the sense of privileged people. Open source often naturally revolves around meritocracy where longer term contributors gain more trust from the developer community. However, some hierarchy is necessary for practical purposes. As such there are repositories “maintainers” who are responsible for merging pull requests as well as a “lead maintainer” who is responsible for overall merging, moderation and appointment of maintainers and the release cycle.
+
+
+Contributor Workflow
+--------------------
+
+The codebase is maintained using the “contributor workflow” where everyone without exception contributes patch proposal using “pull requests”. This facilitates social contribution, easy testing and peer review.
```
Comment (fanquake) [inline CONTRIBUTING.md]: s/proposal/proposals

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+  - Fork repository
+  - Create topic branch
+  - Commit patches 
+
+The project coding conventions in [doc/developer-notes.md](doc/developer-notes.md) must be adhered to.
+
+In general [commits should be atomic](https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention) and diff should be easy to read. For this reason do not mix any formatting fixes or code moves with actual code changes.
```
Comment (fanquake) [inline CONTRIBUTING.md]: s/diff/diffs

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+Commit messages should be verbose by default consisting of a short subject line (50 chars max), a blank line and detailed explanatory text as separate paragraph(s); unless the title alone is self-explanatory (like "fixed spelling") then a single title line is sufficient. Commit messages should be helpful to people reading your code in the future, so explain the reasoning for your decisions. Further explanation [here](http://chris.beams.io/posts/git-commit/).
```
Comment (fanquake) [inline CONTRIBUTING.md]: Should "fixed spelling" be the example we give as a good simple commit message? Maybe "Correct typo in xx.cpp", or something similar?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  - Push changes to your fork
+  - Create pull request
+
+The title of the pull request should be prefixed by the component or area that the pull request affects. Examples:
+
+    [P2P] Add ZMQ message support
+    [Consensus] Add new opcode for BIP-XXXX OP_CHECKAWESOMESIG
+    [Qt Wallet] Refactor send code
+
+If a pull request is specifically not to be considered for merging (yet) please prefix the title with [WIP] or use [Tasks Lists](https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments) in markdown to the body of the pull request to indicate tasks are pending.
```
Comment (fanquake) [inline CONTRIBUTING.md]: s/markdown to the /markdown in the/

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+The body of the pull request should contain enough description about what the patch does together with any justification/reasoning. You should include references to any discussions (for example other tickets or mailing list discussions).
+
+At this stage one should expect comments and review from other contributors. You can add more commits to your pull request by committing them locally and pushing to your fork until you have satisfied all feedback. If your pull request is accepted for merging, you may be asked by a maintainer to squash and or rebase your commits before it will be merged. The length of time for required for peer review is unpredictable and will vary from patch to patch.
```
Comment (fanquake) [inline CONTRIBUTING.md]: s/time for required/time required/

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+
+"Decision Making" Process
+-------------------------
+
+The following applies to code changes to the Bitcoin Core project (and related projects such as secp256k1), and is not to be confused with overall Bitcoin Network Protocol consensus changes.
+
+Whether a pull request is merged into Bitcoin Core rests with the project merge maintainers and ultimately the project lead. 
+
+Maintainers will take into consideration if a patch is in line with the general principles of the project, that the quality of the patch is at the minimum bar and as well as judging the general consensus of contributors.
```
Comment (fanquake) [inline CONTRIBUTING.md]: Suggest rewriting this slightly
*that the quality of the patch meets the minimum standards for inclusion, as well as

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+In general, all pull requests must:
+
+  - have a clear use case, fix a demonstrable bug or serve the greater good of the project (for example refactoring for modularisation);
+  - be well peer reviewed;
+  - have unit tests and functional tests where appropriate;
+  - follow code style guidelines;
+  - not break the existing test suite;
+  - where bugs are fixed, where possible, there should be unit test demonstrating the bug and also proving the fix. This helps prevent regression;
```
Comment (fanquake) [inline CONTRIBUTING.md]: s/be unit test/be a unit test/

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  - where consensus rules are affected, must be accompanied by a published BIP assigned with a number;
+
+###Peer Review
+
+Anyone may participate in peer review is expressed by comments in the pull request. Typically reviewers will review the code for obvious errors, as well as test out the patch set and opine on the technical merits of the patch. Project maintainers take into account the peer review when determining if there is consensus to merge a pull request (remember that discussions may have been spread out over github, mailing list and IRC discussions). The following language is used within pull-request comments:
```
Comment (fanquake) [inline CONTRIBUTING.md]: s/peer review is/peer review which is/

**Comment by fanquake** [comment]
Nice work. Definite improvement on the developer-notes, which were somewhat hidden away. The fact that this is shown as suggested reading to everyone opening a pull request is also good.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
+Contributing to Bitcoin Core
+============================
+
+The Bitcoin Core project operates an open contributor model where anyone is welcome to contribute towards development in the form of peer review, testing and patches. This document explains the practical process and guidelines for contributing.
+
+Firstly in terms of structure, there is no particular concept of “Core developers” in the sense of privileged people. Open source often naturally revolves around meritocracy where longer term contributors gain more trust from the developer community. However, some hierarchy is necessary for practical purposes. As such there are repositories “maintainers” who are responsible for merging pull requests as well as a “lead maintainer” who is responsible for overall merging, moderation and appointment of maintainers and the release cycle.
```
Comment (luke-jr) [inline CONTRIBUTING.md]: s/hierarchy/roles/

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+Please refer to the [Git manual](https://git-scm.com/doc) for more information about Git.
+
+  - Push changes to your fork
+  - Create pull request
+
+The title of the pull request should be prefixed by the component or area that the pull request affects. Examples:
+
+    [P2P] Add ZMQ message support
+    [Consensus] Add new opcode for BIP-XXXX OP_CHECKAWESOMESIG
+    [Qt Wallet] Refactor send code
```
Comment (luke-jr) [inline CONTRIBUTING.md]: Prefer colon format (and seems to be more common):

    p2p: Add ZMQ message support
    consensus: Add new opcode for BIP-XXXX
    qt/wallet: Refactor send code

(also note ZMQ really has nothing to do with p2p..)

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+Pull Request Philosophy
+-----------------------
+
+Patchsets should always be focused. For example, a pull request could add a feature, fix a bug, or refactor code; but not a mixture. Please also avoid super pull requests which attempt to do too much, are overly large, or overly complex as this makes review difficult.
+
+
+###Features
+
+When adding a new feature, thought must be given to the long term technical debt and maintenance that feature may require after inclusion. Before proposing a new feature that will require maintenance, please consider if you are willing to maintain it (including bug fixing). If features get orphaned with no maintainer in the future, they may be removed by the Repository Maintainer.
```
Comment (luke-jr) [inline CONTRIBUTING.md]: Should I reopen TBC support? ;)

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+There are three categories of refactoring, code only moves, code style fixes, code refactoring. In general refactoring pull requests should not mix these three kinds of activity in order to make refactoring pull requests easy to review and uncontroversial. In all cases, refactoring PRs must not change the behaviour of code within the pull request (bugs must be preserved as is).
+
+Project maintainers aim for a quick turnaround on refactoring pull requests, so where possible keep them short, uncomplex and easy to verify. 
+
+
+"Decision Making" Process
+-------------------------
+
+The following applies to code changes to the Bitcoin Core project (and related projects such as secp256k1), and is not to be confused with overall Bitcoin Network Protocol consensus changes.
```
Comment (luke-jr) [inline CONTRIBUTING.md]: libsecp256k1*

Although I'm not sure this document should apply to it.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+Maintainers will take into consideration if a patch is in line with the general principles of the project; meets the minimum standards for inclusion; and will judge the general consensus of contributors.
+
+In general, all pull requests must:
+
+  - have a clear use case, fix a demonstrable bug or serve the greater good of the project (for example refactoring for modularisation);
+  - be well peer reviewed;
+  - have unit tests and functional tests where appropriate;
+  - follow code style guidelines;
+  - not break the existing test suite;
+  - where bugs are fixed, where possible, there should be unit tests demonstrating the bug and also proving the fix. This helps prevent regression;
```
Comment (luke-jr) [inline CONTRIBUTING.md]: This doesn't belong under "must".

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  - where consensus rules are affected, must be accompanied by a published BIP assigned with a number;
```
Comment (luke-jr) [inline CONTRIBUTING.md]: Also needs non-trivial miner support (for new rules) or community consensus (for removed/changed rules).

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+
+###Peer Review
+
+Anyone may participate in peer review which is expressed by comments in the pull request. Typically reviewers will review the code for obvious errors, as well as test out the patch set and opine on the technical merits of the patch. Project maintainers take into account the peer review when determining if there is consensus to merge a pull request (remember that discussions may have been spread out over github, mailing list and IRC discussions). The following language is used within pull-request comments:
+
+  - ACK means "I have tested the code and I agree it should be merged";
+  - NACK means "I disagree this should be merged", and must be accompanied by sound technical justification. NACKs without accompanying reasoning may be disregarded;
```
Comment (luke-jr) [inline CONTRIBUTING.md]: Not all good NACK reasons are technical.

File: CONTRIBUTING.md (hunk 12 above)
Comment (sipa) [inline CONTRIBUTING.md]: s/repositories/repository/

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+  - utACK means "I have not tested the code, but I have reviewed it and it looks OK, I agree it can be merged";
+  - Concept ACK means "I agree in the general principle of this pull request";
+  - Nit refers to trivial, often non-blocking issues.
+
+Project maintainers reserve the right to weigh the opinions of peer reviewers using common sense judgement and also may weight based on meritocracy: Those that have demonstrated a deeper commitment and understanding towards the project (over time) or have clear domain expertise may naturally have more weight, as one would expect in all walks of life.
+
+Where a patch set affects consensus critical code, the bar will be set much higher in terms of discussion and peer review, remembering mistakes could be very costly to the wider community. This includes refactoring of consensus critical code.
```
Comment (instagibbs) [inline CONTRIBUTING.md]: "remembering [that] mistakes" or "keeping in mind that" perhaps?

File: CONTRIBUTING.md (hunk 18 above)
Comment (laanwj) [inline CONTRIBUTING.md]: I'm sure there are some exceptions, but in general they should aim to be.

File: CONTRIBUTING.md (hunk 17 above)
Comment (laanwj) [inline CONTRIBUTING.md]: absolutely. They are a wholly different animal from normal run-of-the-mill changes, that should be made clear.

**Pull request merged**